src/solver/brent/mod.rs

```diff
@@ -0,0 +1,181 @@
+        _state: &IterState<O>,
+    ) -> Result<Option<ArgminIterData<O>>, Error> {
+        self.fa = op.apply(&self.a)?;
+        self.fb = op.apply(&self.b)?;
+        if self.fa * self.fb > 0.0 {
+            return Err(Error::from_boxed_compat(Box::new(BrentError)));
+        }
+        self.fc = self.fb;
+        Ok(Some(
+            ArgminIterData::new().param(self.b).cost(self.fb.abs()),
+        ))
+    }
+
+    fn next_iter(
+        &mut self,
+        op: &mut OpWrapper<O>,
+        // Brent maintains its own state
+        _state: &IterState<O>,
+    ) -> Result<ArgminIterData<O>, Error> {
+        if (self.fb > 0.0 && self.fc > 0.0) || self.fb < 0.0 && self.fc < 0.0 {
+            self.c = self.a;
+            self.fc = self.fa;
+            self.d = self.b - self.a;
+            self.e = self.d;
+        }
+        if self.fc.abs() < self.fb.abs() {
+            self.a = self.b;
+            self.b = self.c;
+            self.c = self.a;
+            self.fa = self.fb;
+            self.fb = self.fc;
+            self.fc = self.fa;
+        }
+        // effective tolerance is double machine precision plus half tolerance as given.
+        let eff_tol = 2.0 * f64::MIN_POSITIVE * self.b.abs() + 0.5 * self.tol;
+        let mid = 0.5 * (self.c - self.b);
+        if mid.abs() <= eff_tol || self.fb == 0.0 {
+            return Ok(ArgminIterData::new()
+                .termination_reason(TerminationReason::TargetPrecisionReached)
+                .param(self.b)
+                .cost(self.fb.abs()));
+        }
+        if self.e.abs() >= eff_tol && self.fa.abs() > self.fb.abs() {
+            let s = self.fb / self.fa;
+            let (mut p, mut q) = if self.a == self.c {
+                (2.0 * mid * s, 1.0 - s)
+            } else {
+                let q = self.fa / self.fc;
+                let r = self.fb / self.fc;
+                (
+                    s * (2.0 * mid * q * (q - r) - (self.b - self.a) * (r - 1.0)),
+                    (q - 1.0) * (r - 1.0) * (s - 1.0),
+                )
+            };
+            if p > 0.0 {
+                q = -q;
+            }
+            p = p.abs();
+            let min1 = 3.0 * mid * q - (eff_tol * q).abs();
+            let min2 = (self.e * q).abs();
+            if 2.0 * p < if min1 < min2 { min1 } else { min2 } {
+                self.e = self.d;
+                self.d = p / q;
+            } else {
+                self.d = mid;
+                self.e = self.d;
+            };
+        } else {
+            self.d = mid;
+            self.e = self.d;
+        };
+        self.a = self.b;
+        self.fa = self.fb;
+        if self.d.abs() > eff_tol {
+            self.b += self.d;
+        } else {
+            self.b += if mid >= 0.0 {
+                eff_tol.abs()
+            } else {
+                -eff_tol.abs()
+            };
+        }
+
+        self.fb = op.apply(&self.b)?;
+        Ok(ArgminIterData::new().param(self.b).cost(self.fb.abs()))
+    }
+}
```

Comment:
Please add a test for the implemented traits, [for instance similar to this](https://github.com/argmin-rs/argmin/blob/5f33e8932c94b010ae62612eb956110d0d0436bf/src/solver/conjugategradient/cg.rs#L144).

Reply:
>You can define custom stopping criteria via the `terminate` method of the `Solver` trait ([see here](https://github.com/argmin-rs/argmin-core/blob/7dc429c0da9592ae9de8ae632454ea3ccc76517a/src/lib.rs#L172)). This will be executed at the [beginning of each iteration](https://github.com/argmin-rs/argmin-core/blob/7dc429c0da9592ae9de8ae632454ea3ccc76517a/src/executor.rs#L148). Alternatively it could also be implemented [here](https://github.com/argmin-rs/argmin-core/blob/7dc429c0da9592ae9de8ae632454ea3ccc76517a/src/lib.rs#L157) which would then apply for all solvers, but I'm not sure if this is appropriate. If I remember correctly, some solvers require to do such checks somewhere in the middle of an iteration, which would then not be possible.

I have already used the possibility to set a custom termination reason, so that should be fine. 
And I agree, controlling the accuracy by the `Executor` might not work in all cases.